Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift

```diff
@@ -0,0 +1,319 @@
+            }
+        }
+
+        /// Initialize a ML-DSA-65 private key from a PEM representation.
+        /// 
+        /// - Parameter pemRepresentation: The PEM representation of the private key.
+        public init(pemRepresentation: String) throws {
+            let document = try ASN1.PEMDocument(pemString: pemRepresentation)
+            self = try .init(derRepresentation: document.derBytes)
+        }
+
+        /// The public key associated with this private key.
+        public var publicKey: PublicKey {
+            get throws {
+                try PublicKey(privateKey: self)
+            }
+        }
+
+        /// Generate a signature for the given data.
+        /// 
+        /// - Parameters: 
+        ///   - data: The message to sign.
+        ///   - context: The context to use for the signature.
+        /// 
+        /// - Returns: The signature of the message.
+        public func signature(for data: some DataProtocol, context: [UInt8]? = nil) throws -> Signature {
+            let output = try Array<UInt8>(unsafeUninitializedCapacity: Signature.bytesCount) { bufferPtr, length in
+                let result = data.regions.first!.withUnsafeBytes { dataPtr in
+                    let contextPointer = context?.withUnsafeBufferPointer { $0.baseAddress }
+                    let contextLength = context?.count ?? 0
+
+                    return CCryptoBoringSSL_MLDSA65_sign(
+                        bufferPtr.baseAddress,
+                        self.pointer,
+                        dataPtr.baseAddress,
+                        dataPtr.count,
+                        contextPointer,
+                        contextLength
+                    )
+                }
+
+                guard result == 1 else {
+                    throw CryptoMLDSAError.signatureGenerationFailure
+                }
+
+                length = Signature.bytesCount
+            }
+            return Signature(signatureBytes: output)
+        }
+
+        /// The size of the private key in bytes.
+        public static let bytesCount = 4032
+    }
+}
+
+extension MLDSA {
+    /// A ML-DSA-65 public key.
+    public struct PublicKey: Sendable {
+        private let pointer: UnsafeMutablePointer<MLDSA65_public_key>
+
+        fileprivate init(privateKey: PrivateKey) throws {
+            self.pointer = UnsafeMutablePointer<MLDSA65_public_key>.allocate(capacity: 1)
+
+            guard CCryptoBoringSSL_MLDSA65_public_from_private(
+                self.pointer,
+                privateKey.pointer
+            ) == 1 else {
+                throw CryptoMLDSAError.keyGenerationFailure
+            }
+        }
+
+        /// Initialize a ML-DSA-65 public key from a seed.
+        /// 
+        /// The seed must be at least 32 bytes long.
+        /// Any additional bytes in the seed are ignored.
+        /// 
+        /// - Parameter seed: The seed to use to generate the public key.
+        public init(from seed: some DataProtocol) throws {
```

Comment:
I don't think this API makes sense: users should construct the private key, not the public one.